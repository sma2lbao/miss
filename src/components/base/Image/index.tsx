import * as React from "react";
import { makeStyles, createStyles, Theme } from "@material-ui/core/styles";
import { Skeleton } from "@material-ui/lab";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      height: (props: ImageProp) => {
        return props.height || "auto";
      },
      width: (props: ImageProp) => {
        return props.width || "100%";
      },
      fontSize: 0,
      overflow: "hidden"
    },
    aspect: {
      paddingTop: (props: ImageProp) => {
        if (props.aspectRatio) {
          return `calc(1 / ${props.aspectRatio} * 100%)`;
        } else {
          return 0;
        }
      },
      position: "relative",
      backgroundColor: "transparent"
    },
    image: {
      position: (props: ImageProp) => {
        if (props.aspectRatio) {
          return "absolute";
        } else {
          return "relative";
        }
      },
      width: "100%",
      height: "100%",
      top: 0,
      left: 0,
      objectFit: "cover"
    },
    container: {
      position: "absolute",
      width: "100%",
      height: "100%",
      top: 0,
      left: 0,
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      pointerEvents: "none"
    }
  })
);

interface ImageProp {
  aspectRatio?: number; // 宽高比
  src?: string;
  classes?: Object;
  width?: number | string;
  height?: number | string;
  alt?: string;
}

enum ImageStatus {
  Loading = 0,
  Success = 1,
  Failuare = -1
}

export const Image: React.FC<ImageProp> = (props: ImageProp) => {
  const classes = useStyles(props);
  const { src, alt } = props;

  // status: 0:加载中 1: 加载成功 -1:加载失败
  const [status, setStatus] = React.useState(ImageStatus.Loading);

  const _loadImage = () => {
    setStatus(ImageStatus.Success);
  };

  const _loadImageFail = () => {
    setStatus(ImageStatus.Failuare);
  };

  return (
    <div className={classes.root}>
      {props.aspectRatio && (
        <div className={classes.aspect}>
          {src && status !== ImageStatus.Failuare && (
            <img
              alt={alt}
              className={classes.image}
              src={src}
              onLoad={_loadImage}
              onError={_loadImageFail}
            />
          )}
          {status !== ImageStatus.Success && (
            <div className={classes.container}>
              {status === ImageStatus.Loading && (
                <Skeleton
                  variant="rect"
                  animation="wave"
                  height="100%"
                  width="100%"
                />
              )}
              {status === ImageStatus.Failuare && (
                <Skeleton
                  variant="rect"
                  animation="wave"
                  height="100%"
                  width="100%"
                />
              )}
            </div>
          )}
        </div>
      )}
      {!props.aspectRatio && (
        <>
          {src && status !== ImageStatus.Failuare && (
            <img
              className={classes.image}
              src={src}
              alt=""
              onLoad={_loadImage}
              onError={_loadImageFail}
            />
          )}
          {status !== ImageStatus.Success && (
            <div className={classes.container}>
              {status === ImageStatus.Loading && (
                <Skeleton
                  variant="rect"
                  animation="wave"
                  height="100%"
                  width="100%"
                />
              )}
              {status === ImageStatus.Failuare && (
                <Skeleton
                  variant="rect"
                  animation="wave"
                  height="100%"
                  width="100%"
                />
              )}
            </div>
          )}
        </>
      )}
    </div>
  );
};

export default Image;
