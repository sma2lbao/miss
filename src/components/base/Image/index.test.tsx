import Image from "./index";
import { render } from "enzyme";
import * as React from "react";

it("renders image", () => {
  const component = render(<Image />);
  expect(component).toMatchSnapshot();
});
