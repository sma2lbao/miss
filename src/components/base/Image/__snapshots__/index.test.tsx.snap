// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders image 1`] = `
<div
  class="makeStyles-root-1 makeStyles-root-5"
>
  <div
    class="makeStyles-container-4"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-wave"
      style="width:100%;height:100%"
    />
  </div>
</div>
`;
